prow/cmd/image-builder/buildcontroller.go

```diff
@@ -253,110 +266,50 @@ func (r *buildReconciler) createPVC(ctx context.Context, ibPod *corev1.Pod) (*co
-			},
-			Spec: corev1.PodSpec{
-				RestartPolicy: corev1.RestartPolicyNever,
-				Volumes: []corev1.Volume{
-					{
-						Name: dockerConfigVolume,
-						VolumeSource: corev1.VolumeSource{
-							Secret: &corev1.SecretVolumeSource{
-								SecretName: r.options.dockerConfigSecret,
-							},
-						},
-					},
-				},
-			},
-		}
+	var variants []buildVariant
 
-		// Base configuration of kaniko container
-		kanikoContainer := corev1.Container{
-			Name:  "kaniko",
-			Image: r.options.kanikoImage,
-			Args: []string{
-				"--skip-unused-stages",
-				"--context=/code",
-				fmt.Sprintf("--dockerfile=%s", r.options.dockerfile),
-				fmt.Sprintf("--target=%s", target),
-			},
-			VolumeMounts: []corev1.VolumeMount{
-				{
-					Name:      codeVolume,
-					MountPath: "/code",
-					SubPath:   "code",
-				},
-				{
-					Name:      dockerConfigVolume,
-					MountPath: "/kaniko/.docker",
-				},
-			},
-			Resources: corev1.ResourceRequirements{
-				Requests: corev1.ResourceList{
-					corev1.ResourceCPU:    qtyZero,
-					corev1.ResourceMemory: qtyZero,
-				},
-			},
-		}
-
-		// Add destinations
-		destinations, err := r.defineDestinations(target)
+	if r.options.buildDefintion != "" {
+		// Get variants for build-definition case
+		variants, err = r.getVariants()
 		if err != nil {
-			return errors.Wrap(err, "construct destinations")
-		}
-		kanikoContainer.Args = append(kanikoContainer.Args, destinations...)
-
-		// Add kaniko args
-		kanikoContainer.Args = append(kanikoContainer.Args, r.options.kanikoArgs.Strings()...)
-
-		// Set caching options
-		if r.options.cacheRegistry != "" {
-			kanikoContainer.Args = append(
-				kanikoContainer.Args,
-				"--cache=true",
-				fmt.Sprintf("--cache-repo=%s", r.options.cacheRegistry),
-			)
+			return errors.Wrap(err, "get build variants")
 		}
+		r.log.Infof("Selected variants for this build: %+v", variants)
+	} else {
+		// Empty variant for the regular build case
+		variants = append(variants, buildVariant{name: nil, buildArgs: nil})
+	}
 
-		if i == 0 && r.options.cacheRegistry != "" {
-			buildGroup = "createCache"
-		} else {
-			buildGroup = "parallelBuild"
-		}
+	// Next pods build the targets for the variants
+	for _, variant := range variants {
+		for i, target := range r.options.targets.Strings() {
 
-		// Append build container to build pod
-		pod.Spec.Containers = append(pod.Spec.Containers, kanikoContainer)
+			pod, err := r.definePodForTarget(ibPod, target, variant.name, variant.buildArgs)
+			if err != nil {
+				return errors.Wrapf(err, "define build pod for target %s", target)
+			}
 
-		// Configure the build pod with PVC, node assignment and controller reference
-		r.assignPVC(&pod)
-		r.setNodeAssignment(ibPod, &pod)
-		err = controllerutil.SetControllerReference(ibPod, &pod, r.scheme)
-		if err != nil {
-			return errors.Wrap(err, "set controller reference")
-		}
+			// Set build group
+			var buildGroup string
+			if i == 0 && r.options.cacheRegistry != "" {
```

Comment:
In `startNextBuildPods` you assume that `r.buildPods` is sorted by build group (i.e. `clonerefs`, then all `createCache` and then all `parallel`).
With multiple targets and build variants, this is no longer the case, right?

Reply:
When cache is enabled and there are multiple variants the groups and their sequence will look like this
- **1** "clonerefs" 
- **2** "createCache-variant1"
- **3** "parallelBuild" (for variant 1)
- **3** "parallelBuild" (for variant 1)
- **3**  ... (for variant 1)
- **4** "createCache-variant2"
- **5** "parallelBuild" (for variant 2)
- **5** ... (for variant 1)

The slice is not sorted after it was created, so it will always build one variant after another. Thus, variant 2 could use cached layers created by variant 1.

When you disable cache everything except the cloning will run in parallel
- **1** "clonerefs"
- **2** "parallelBuild" (for variant 1)
- **2** ... (for variant 1)
- **2** "parallelBuild" (for variant 2)
- **2** ... (for variant 2)

The reason for adding variant name to the build group is when have multiple variants of a single stage Dockerfile and you enable cache explicitly
- **1** "clonerefs" 
- **2** "createCache-variant1"
- **3** "createCache-variant2"
In this case variant 2 would be able to consume the cache created by variant 1 as in the multi stage case.

Reply:
Alrighty. Then this seems to be what we want. I wasn't sure, if this is by intent, but makes sense :)